Shell/main.cpp

```diff
@@ -847,6 +1060,35 @@ static ExitCodeOrContinuationRequest::ContinuationRequest is_complete(const Vect
     return ExitCodeOrContinuationRequest::Nothing;
 }
 
+static IterationDecision wait_for_pid(pid_t pid, const String& name, bool is_first_command_in_chain, int& return_value)
+{
+    int wstatus = 0;
+    int rc = waitpid(pid, &wstatus, WSTOPPED);
+    if (rc < 0 && errno != EINTR) {
+        if (errno != ECHILD)
+            perror("waitpid");
+        return IterationDecision::Break;
+    }
+    auto job_id = g.jobs.get(pid).value_or(Job {}).job_id();
+    if (WIFEXITED(wstatus)) {
+        if (WEXITSTATUS(wstatus) != 0)
+            dbg() << "Shell: " << name << ":" << pid << " exited with status " << WEXITSTATUS(wstatus);
+        if (is_first_command_in_chain)
+            return_value = WEXITSTATUS(wstatus);
+        g.jobs.remove(pid);
+    } else if (WIFSTOPPED(wstatus)) {
+        fprintf(stderr, "Shell: [%llu] %s(%d) %s\n", job_id, name.characters(), pid, strsignal(WSTOPSIG(wstatus)));
+    } else {
+        if (WIFSIGNALED(wstatus)) {
+            printf("Shell: [%llu] %s(%d) exited due to signal '%s'\n", job_id, name.characters(), pid, strsignal(WTERMSIG(wstatus)));
+        } else {
+            printf("Shell: [%llu] %s(%d) exited abnormally\n", job_id, name.characters(), pid);
+        }
```

Comment:
And here — again, for the future — instead of emitting boring text, emit some magic escape characters to let the terminal draw the card red and display the exit code / signal number elsewhere.

Reply:
This needs a lot of thought and discussion, I am more in favour of doing this via RPC, but kling is also very right in that it should work for remote hosts as well
we could certainly add Serenity Terminal-specific escape sequences, but it just turns into another mess on top of more previous messes

Needless to say, I am very much willing to do this, once we settle on something.